### ISSUE TITLE
XrmQuery: Proper encoding of query string parameters

### ISSUE THREAD

**Issue by bo-stig-christensen** [delegateas/XrmDefinitelyTyped#59]
Hi,

There is an issue with _XrmQuery.sendRequest()_ in _dg.xrmquery.web.js_. - The fragment in question is:
```
  export function sendRequest(type: XQW.HttpRequestType, queryString: string, data: any, successCb: (x: XMLHttpRequest) => any, errorCb?: (err: Error) => any, configure?: (req: XMLHttpRequest) => void): void {
    request(type, encodeURI(XQW.getApiUrl() + queryString), data, successCb, errorCb, configure);
  }
```

If you trigger a request containing a query string with characters not escaped by encodeURI(), the request will fail. - An example is if one of the query string parameters contain e.g. "C++" or "C#". The characters '+' and '#' are preserved with encodeURI().
Instead, it is necessary to encode each query string parameter individually using encodeURIComponent().
As an example, I have created a _very_ basic UrlEncoder that takes this into consideration:
```
namespace UrlEncoder
{
    /**
     * Encode url with query string and each query parameter individually encoded.
     * @author Bo Stig Christensen
     * @date 20.02.2018
     * @param {string} url Absolute URL to encode.
     * @return {string} Encoded URL.
     */
    export function encodeWithQueryString(url: string): string {
        if (url == null || url.indexOf("?") === -1)
            return url;

        let urlComponents = url.split("?");
        let queryString = urlComponents[1];

        let queries = queryString.split("&");

        let encodedQueryParams: Array<string> = new Array<string>();
        queries.forEach((queryParam) => {
            let keyValuePair = queryParam.split("=");
            keyValuePair[0] = encodeURIComponent(keyValuePair[0]);
            keyValuePair[1] = encodeURIComponent(keyValuePair[1]);
            encodedQueryParams.push(keyValuePair.join("="));
        });

        urlComponents[0] = encodeURI(urlComponents[0]);
        urlComponents[1] = encodedQueryParams.join("&");
        return urlComponents.join("?");
    }
}
```

This would then be used as follows in XrmQuery.sendRequest() function:
```
request(type, UrlEncoder.encodeWithQueryString(XQW.getApiUrl() + queryString), data, successCb, errorCb, configure);
```

The question is if you consider this encoding the responsibility of the client or the XrmQuery module? - It would of course make life easier for the user of XrmQuery if it handles the encoding internally :-)

**Comment by majakubowski** [comment]
Function proposed by Bo doesn't resolve problem. Function can be easy broken by string: "=+". I think the proper way is to encode every parameter provided by developer. I assume that queryString will be properly escaped will be passed directly to XrmQuery.sendRequest.

**Comment by bo-stig-christensen** [comment]
@majakubowski I think the solution you implemented does not take into consideration that sendRequest() (and even request()) is exported and therefore publicly available. By only encoding at the XrmQuery.Filter level there are still public functions available deeper down the hierarchy, which would then not be properly encoded.

The safest place to fix the issue is at its root, namely in the lowest level request(). - Unless the author decides that encoding is the responsibility of the client. But it could be great to have this functionality directly in the library to "not have to worry" ;-)

My above "UrlEncoder" was just a simple example, but since I have not used that much energy on it, I would simply recommend to find an already proven encoder on npm or something. - Usually, someone with a dedicated library will have spent more energy on "getting things right".

That said, I just did a fix to my example above that will fix the issues you identified, as well as add a multi-encoding safety-check that will work 99.9% of the time. Only case that won't work is if you actually want to untouched pass in something like "%20" or "%25" or another "%" followed by 2 digits. However, the chances are hypothetical I guess?

```
namespace UrlEncoder
{
    /**
     * Encode url with query string and each query parameter individually encoded.
     * @author Bo Stig Christensen
     * @date 22.02.2018
     * @param {string} url Absolute URL to encode.
     * @return {string} Encoded URL.
     */
    export function encodeUrlWithQueryString(url: string): string {
        // Baic sanity null/empty check.
        if (url == null || url.trim().length === 0)
            return url;

        // Ensure that url is decoded (this approach will work 99.9% of the time).
        while (url.search(new RegExp("%[0-9]{2}")) !== -1)
            url = decodeURI(url);

        // Simply return at this point if no query string is present.
        if (url.indexOf("?") === -1)
            return url;

        // Encode individual query string parameters.
        let urlComponents = url.split("?");
        urlComponents[0] = encodeURI(urlComponents[0]);
        urlComponents[1] = encodeQueryString(urlComponents[1]);

        return urlComponents.join("?");
    }

    export function encodeQueryString(queryString: string): string {
        let queries = queryString.split("&");

        // Encode individual query string parameters.
        let encodedQueryParams = new Array<string>();
        queries.forEach((queryParam) => {
            let keyValuePair = queryParam.split("=");
            let key = encodeURIComponent(keyValuePair[0]);
            let value = "";
            if (keyValuePair.length > 1) {
                value = keyValuePair.slice(1).join("=");
                value = "=" + encodeURIComponent(value);
            }
            encodedQueryParams.push(key + value);
        });
        let encodedQueryString = encodedQueryParams.join("&");

        // 2nd pass to parse remaining '&' that are not actual parameter separators.
        // 
        // NOTE: The only scenario not covered properly is using empty query params like: "myFirstParam=123&myEmptyParam&myOtherParam=456".
        //       This should however not happen with CRM's API and XrmQuery's implementation.
        let splitParams = encodedQueryString.split("=");
        let encodedSplitParams = new Array<string>();
        splitParams.forEach((splitParam, index) => {
            let splitParamSections = splitParam.split("&");
            let value = splitParamSections[0];
            let nextParamName = "";
            if (splitParamSections.length > 1) {
                if (index < splitParams.length - 1)
                    nextParamName = splitParamSections.pop() as string;
                value = splitParamSections.join("%23"); // '%23' is '&' encoded.
                if (nextParamName)
                    value += "&";
            }
            encodedSplitParams.push(value + nextParamName);
        });
        encodedQueryString = encodedSplitParams.join("=");

        return encodedQueryString;
    }
}
```

Again, this is not a fully-fledged UrlEncoder (and has a couple of quirks), for which I guess there has to be a better library out there. the above should just be considered a working sample (and possible solution until something better comes along).

@mktange What do you think? :-)

**Comment by mktange** [comment]
I am not a fan of adding a lot of code for handling this properly for all cases (i.e. include/embed a third-party library). 

I'm more leaning towards something like what @majakubowski has done, where XrmQuery makes sure that the XrmQuery-specific parts of the query string are encoded correctly. IMO it is not the responsibility of XrmQuery to handle every arbitrary requests a developer may throw at `sendRequest()` and `request()`. Those low level functions are to be used "at your own risk".

**Comment by TomMalow** [comment]
I have discussed it with the other guys and we also agree with @mktange.

First of, `sendrequest()` is not in the spirit of XDT and is only there to cover anything that XDT cannot handle as mentioned by @mktange. And if there is such a case, then XDT should be updated to handle that case.
Second, if a developer is going to use `sendRequest()` then it is that developers responsibility to provide a valid query string. It is not in our interest to validate the input on a method that we don't believe a developer should use in the first place. Especially if a better method exist in XDT. This also means that it is XDT jobs to encode anything not related to the query string as suggested by @majakubowski's pull request.

I believe we will go with @majakubowski change request but we do have some slight modification that we will add. 

Addtionally, we are considering changing `request()` method to not be public. It is even tagged as internal. But we are not sure if there are any major issues in limiting the access to this function.

**Comment by majakubowski** [comment]
After I discovered unit tests in XrmDefinitelyTyped I reviewed my changes one more time.. :)
And after reading few threads at stackoverflow or similar about odata encoding I decided to make custom function for encoding parameters. The reason is simple - encoding rules is far more complex than simple encoding each entrie parameter ($select/$filter/$expand) or entrie query string. The simplest example - single quote should be replaced with two single quotes - http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/abnf/odata-abnf-construction-rules.txt :

> One prominent example is the single quote that delimits OData primitive type literals: **%27 and ' are treated identically**, so **a single quote within a string literal is "encoded" as two consecutive single quotes** in either literal or percent-encoded representation.

I added few unit test to cover special character cases both in query string (getQueryString method) and in http request (to make sure that query strings are not double encoded).
